pkg/cdn/download.go

```diff
@@ -1,13 +1,93 @@
 package cdn
 
 import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+
+	"github.com/gomods/athens/pkg/config"
+	"github.com/gomods/athens/pkg/config/env"
 	"github.com/gomods/athens/pkg/storage"
 )
 
+// ModVerDownloader downloads a module version from a URL
+type ModVerDownloader func(ctx context.Context, baseURL, module, version string) (*storage.Version, error)
+
 // Download downloads the module/version from url. Returns a storage.Version
 // representing the downloaded module/version or a non-nil error if something
 // went wrong
-func Download(url, module, version string) (*storage.Version, error) {
-	// TODO
-	return nil, nil
+func Download(ctx context.Context, baseURL, module, version string) (*storage.Version, error) {
+	tctx, cancel := context.WithTimeout(ctx, env.Timeout())
+	defer cancel()
+	getReq := func(ext string) (*http.Request, error) {
+		return getRequest(tctx, baseURL, module, version, ext)
+	}
+
+	infoReq, err := getReq(".info")
+	if err != nil {
+		return nil, err
+	}
+	modReq, err := getReq(".mod")
+	if err != nil {
+		return nil, err
+	}
+	zipReq, err := getReq(".zip")
+	if err != nil {
+		return nil, err
+	}
+
+	info, err := getResBytes(infoReq)
+	if err != nil {
+		return nil, err
+	}
+	mod, err := getResBytes(modReq)
+	if err != nil {
+		return nil, err
+	}
+	zipRes, err := http.DefaultClient.Do(zipReq)
+	if err != nil {
+		return nil, err
+	}
+
+	ver := storage.Version{
+		Info: info,
+		Mod:  mod,
+		Zip:  zipRes.Body,
+	}
+	return &ver, nil
+}
+
+func getResBytes(req *http.Request) ([]byte, error) {
+	res, err := http.DefaultClient.Do(req)
```

Comment:
careful about defaultClient it does not have timeout set up and can hang. 
if we stick with ctx.WithTimeout this should not be a problem, but we can set Timeout also here just to make sure we work if somebody removes WithTimeout